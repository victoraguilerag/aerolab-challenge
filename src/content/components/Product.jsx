import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Loader from 'react-loader-spinner'
import affortableIcon from '../../icons/buy-blue.svg'
import affortableSelectedIcon from '../../icons/buy-white.svg'
import coin from '../../icons/coin.svg'

class Product extends Component {
  state = {
    loading: true
  }
  showImage = (e) => {
    this.setState({
      loading: false
    })
  }
  render () {
    const {
      coins,
      selected,
      info,
      handleProductClick,
      handleRedeemProduct,
      handleModal
    } = this.props
    const id = info.get('_id')
    const name = info.get('name')
    const category = info.get('category')
    const url = info.get('img').get('url')
    const hdUrl = info.get('img').get('hdUrl')
    const cost = info.get('cost')

    const affortable = (coins >= cost)
    return (
      <div
        className="product-card"
        style={styles.productCard}
        onClick={
          (affortable)
            ? (e) => handleProductClick(id) : (e) => handleModal('points')
        }
      >
        {
          this.state.loading &&
          <div className="loader" style={styles.loader}>
            <Loader
              type="TailSpin"
              color="#00bfff"
              height="50"
              width="100"
            />
          </div>
        }
        <img src={url} srcSet={`${hdUrl} 2x`} alt={name} style={styles.productImage} onLoad={() => this.showImage()} onError={() => this.showImage()} />
        {
          !affortable &&
          <div className="productUnavailable" style={styles.productUnavailable}>
            <p className="productUnavailableText" style={styles.productUnavailableText}>{`You need ${cost}`}</p>
            <img
              className="productUnavailableCoin"
              style={styles.productUnavailableCoin}
              width="24px"
              height="24px"
              src={coin}
              alt="coin"
            />
          </div>
        }
        {
          !selected && affortable &&
          <img
            src={affortableIcon}
            alt="affortable"
            style={styles.topIcon}
            width="42px"
            height="42px"
          />
        }
        <div className="product-info" style={styles.productInfo}>
          <div className="product-category" style={styles.productCategory}>
            {category}
          </div>
          <div className="product-name" style={styles.productName}>
            {name}
          </div>
        </div>
        {
          selected &&
          <div className="collapsable-buy" style={styles.collapsableBuy}>
            <img
              src={affortableSelectedIcon}
              alt="affortable"
              style={styles.topIcon}
              width="42px"
              height="42px"
            />
            <div className="product-price" style={styles.productPrice}>
              {cost}
              <img
                className="coin"
                style={styles.coin}
                width="24px"
                height="24px"
                src={coin}
                alt="coin"
              />
            </div>
            <div className="reedem-buttom" style={styles.reedemButton} onClick={() => handleRedeemProduct(id)} >
              <p className="reedem-text" style={styles.reedemText}>
                Reedem now
              </p>
            </div>
          </div>
        }
      </div>
    )
  }
}

const styles = {
  'productCard': {
    'userSelect': 'none',
    'background': '#ffffff',
    'boxShadow': '2px 2px 4px 0 rgba(0,0,0,0.10)',
    'width': '276px',
    'height': '276px',
    'position': 'relative',
    'marginTop': '41px',
    'cursor': 'pointer'
  },
  'productImage': {
    'width': '252px',
    'height': '182px',
    'marginTop': '12px',
    'pointerEvents': 'none'
  },
  'topIcon': {
    'position': 'absolute',
    'top': '12px',
    'right': '12px',
    'pointerEvents': 'none'
  },
  'productInfo': {
    'position': 'absolute',
    'bottom': '15px',
    'left': '24px',
    'pointerEvents': 'none'
  },
  'productCategory': {
    'fontSize': '16px',
    'color': '#a3a3a3',
    'letterSpacing': '-0.04px',
    'textAlign': 'left',
    'pointerEvents': 'none'
  },
  'productName': {
    'fontSize': '18px',
    'color': '#616161',
    'letterSpacing': '-0.04px',
    'textAlign': 'left',
    'pointerEvents': 'none'
  },
  'collapsableBuy': {
    'display': 'flex',
    'flexFlow': 'column',
    'alignItems': 'center',
    'position': 'absolute',
    'top': '0',
    'left': '0',
    'width': '100%',
    'height': '100%',
    'opacity': '1',
    'backgroundImage': 'linear-gradient(-180deg, rgba(10, 212, 250, .6) 0%, rgba(37, 187, 241, .7) 100%)'
  },
  'productPrice': {
    'display': 'flex',
    'flexFlow': 'row wrap',
    'justifyContent': 'center',
    'alignItems': 'center',
    'fontSize': '36px',
    'color': '#ffffff',
    'letterSpacing': '-0.08px',
    'textAlign': 'center',
    'marginTop': '99px',
    'marginBottom': '12px'
  },
  'coin': {
    'width': '26px',
    'height': '26px',
    'margin': '0 5px'
  },
  'reedemButton': {
    'background': '#ffffff',
    'width': '228px',
    'height': '42px',
    'margin': '0 auto',
    'borderRadius': '100px',
    'display': 'flex',
    'justifyContent': 'center',
    'alignItems': 'center'
  },
  'reedemText': {
    'fontSize': '18px',
    'color': '#616161',
    'letterSpacing': '-0.04px',
    'textAlign': 'center'
  },
  'productUnavailable': {
    'position': 'absolute',
    'top': '12px',
    'right': '12px',
    'pointerEvents': 'none',
    'display': 'flex',
    'justifyContent': 'center',
    'alignItems': 'center',
    'opacity': '0.8',
    'background': '#616161',
    'borderRadius': '100px',
    'width': '142px',
    'height': '42px'
  },
  'productUnavailableText': {
    'fontSize': '14px',
    'color': '#ffffff',
    'letterSpacing': '-0.03px',
    'textAlign': 'right'
  },
  'productUnavailableCoin': {
    'width': '20px',
    'height': '20px',
    'margin': '0 2.5px'
  },
  'loader': {
    'width': '252px',
    'height': '182px',
    'display': 'flex',
    'alignItems': 'center',
    'justifyContent': 'center',
    'marginTop': '12px'
  }
}

Product.propTypes = {
  coins: PropTypes.number.isRequired,
  selected: PropTypes.bool.isRequired,
  info: PropTypes.shape({}).isRequired,
  handleRedeemProduct: PropTypes.func.isRequired,
  handleProductClick: PropTypes.func.isRequired,
  handleModal: PropTypes.func.isRequired
}

export default Product
