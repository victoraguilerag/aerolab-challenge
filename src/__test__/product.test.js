import React from 'react'
import Product from '../content/components/Product.jsx'
import productsStub from '../stub/products.js'
import renderer from 'react-test-renderer'

const coins = 2000
const selected = true
const info = productsStub[0]
const functionMock = () => console.log('mock')

test('Product renders', () => {
  const component = renderer.create(
    <Product
      coins={coins}
      selected={selected}
      info={info}
      handleRedeemProduct={functionMock}
      handleModal={functionMock}
      handleProductClick={functionMock}
    />,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
