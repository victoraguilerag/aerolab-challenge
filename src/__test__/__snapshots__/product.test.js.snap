// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders 1`] = `
<div
  className="product-card"
  onClick={[Function]}
  style={
    Object {
      "background": "#ffffff",
      "boxShadow": "2px 2px 4px 0 rgba(0,0,0,0.10)",
      "cursor": "pointer",
      "height": "276px",
      "marginTop": "41px",
      "position": "relative",
      "userSelect": "none",
      "width": "276px",
    }
  }
>
  <div
    className="loader"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "182px",
        "justifyContent": "center",
        "marginTop": "12px",
        "width": "252px",
      }
    }
  >
    <div>
      <svg
        height="50"
        viewBox="0 0 38 38"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <linearGradient
            id="a"
            x1="8.042%"
            x2="65.682%"
            y1="0%"
            y2="23.865%"
          >
            <stop
              offset="0%"
              stopColor="#00bfff"
              stopOpacity="0"
            />
            <stop
              offset="63.146%"
              stopColor="#00bfff"
              stopOpacity=".631"
            />
            <stop
              offset="100%"
              stopColor="#00bfff"
            />
          </linearGradient>
        </defs>
        <g
          fill="none"
          fillRule="evenodd"
        >
          <g
            transform="translate(1 1)"
          >
            <path
              d="M36 18c0-9.94-8.06-18-18-18"
              id="Oval-2"
              stroke="url(#a)"
              strokeWidth="2"
            >
              <animateTransform
                attributeName="transform"
                dur="0.9s"
                from="0 18 18"
                repeatCount="indefinite"
                to="360 18 18"
                type="rotate"
              />
            </path>
            <circle
              cx="36"
              cy="18"
              fill="#fff"
              r="1"
            >
              <animateTransform
                attributeName="transform"
                dur="0.9s"
                from="0 18 18"
                repeatCount="indefinite"
                to="360 18 18"
                type="rotate"
              />
            </circle>
          </g>
        </g>
      </svg>
    </div>
  </div>
  <img
    alt="Secador de Cabello"
    onError={[Function]}
    onLoad={[Function]}
    src="asdas"
    srcSet="asdas 2x"
    style={
      Object {
        "height": "182px",
        "marginTop": "12px",
        "pointerEvents": "none",
        "width": "252px",
      }
    }
  />
  <div
    className="product-info"
    style={
      Object {
        "bottom": "15px",
        "left": "24px",
        "pointerEvents": "none",
        "position": "absolute",
      }
    }
  >
    <div
      className="product-category"
      style={
        Object {
          "color": "#a3a3a3",
          "fontSize": "16px",
          "letterSpacing": "-0.04px",
          "pointerEvents": "none",
          "textAlign": "left",
        }
      }
    >
      electrodomestico
    </div>
    <div
      className="product-name"
      style={
        Object {
          "color": "#616161",
          "fontSize": "18px",
          "letterSpacing": "-0.04px",
          "pointerEvents": "none",
          "textAlign": "left",
        }
      }
    >
      Secador de Cabello
    </div>
  </div>
  <div
    className="collapsable-buy"
    style={
      Object {
        "alignItems": "center",
        "backgroundImage": "linear-gradient(-180deg, rgba(10, 212, 250, .6) 0%, rgba(37, 187, 241, .7) 100%)",
        "display": "flex",
        "flexFlow": "column",
        "height": "100%",
        "left": "0",
        "opacity": "1",
        "position": "absolute",
        "top": "0",
        "width": "100%",
      }
    }
  >
    <img
      alt="affortable"
      height="42px"
      src={Object {}}
      style={
        Object {
          "pointerEvents": "none",
          "position": "absolute",
          "right": "12px",
          "top": "12px",
        }
      }
      width="42px"
    />
    <div
      className="product-price"
      style={
        Object {
          "alignItems": "center",
          "color": "#ffffff",
          "display": "flex",
          "flexFlow": "row wrap",
          "fontSize": "36px",
          "justifyContent": "center",
          "letterSpacing": "-0.08px",
          "marginBottom": "12px",
          "marginTop": "99px",
          "textAlign": "center",
        }
      }
    >
      2000
      <img
        alt="coin"
        className="coin"
        height="24px"
        src={Object {}}
        style={
          Object {
            "height": "26px",
            "margin": "0 5px",
            "width": "26px",
          }
        }
        width="24px"
      />
    </div>
    <div
      className="reedem-buttom"
      onClick={[Function]}
      style={
        Object {
          "alignItems": "center",
          "background": "#ffffff",
          "borderRadius": "100px",
          "display": "flex",
          "height": "42px",
          "justifyContent": "center",
          "margin": "0 auto",
          "width": "228px",
        }
      }
    >
      <p
        className="reedem-text"
        style={
          Object {
            "color": "#616161",
            "fontSize": "18px",
            "letterSpacing": "-0.04px",
            "textAlign": "center",
          }
        }
      >
        Reedem now
      </p>
    </div>
  </div>
</div>
`;
